Param requirements

Blacksmoke16 opened the pull request:
* Run crystal tool format
* Implement Ameba

Blacksmoke16 committed:
path argument requirements

Run crystal tool format
Implement ameba

drujensen reviewed (approved):
💯

elorest reviewed (approved):
That looks good.

robacarp commented on README.md:
```diff
@@ -49,45 +58,45 @@ result.params #=> { "post_name" => "my_trip_to_kansas" }
 
 /put/products/Winter-Windproof-Trapper-Hat/dp/B01J7DAMCQ
-globs with suffix match 667.91k (   1.5µs) (± 3.89%) fastest
+globs with suffix match   1.12M (896.18ns) (± 2.20%)  1584 B/op  fastest
```
I’d love to see an added metric for the filters you’re adding here

robacarp commented on README.md:
```diff
@@ -49,45 +58,45 @@ result.params #=> { "post_name" => "my_trip_to_kansas" }
 /get/books/23/pages
-router: wrong   1.42M (702.46ns) (± 3.84%)       fastest
- radix: wrong   1.01M (991.22ns) (± 1.30%)  1.41× slower
+router: wrong   2.36M (423.88ns) (± 2.74%)  768 B/op        fastest
```
It’s interesting to see the speed difference here, which I’m assuming is due to change in crystal versions

Blacksmoke16 replied:
Either that or I have a beefier PC than what ran it previously.

robacarp commented:
I have a couple thoughts, but I like the idea. I apparently can’t submit my comments on mobile... later I guess. Nice work though. 

Running ameba while also introducing changes makes it a bit hard to see he changes here. Would you mind breaking that out into a different pull?

Blacksmoke16 commented:
@robacarp Sure, can wait till you get a proper review in.

Most of the changes from ameba are specs.

Relevant files for this change would be:
* route_set.cr
* segment.cr
* routes_with_variables_spec.cr

Rest are ameba related only.

Sija commented on src/amber/router/segment.cr:
```diff
@@ -5,8 +5,9 @@ module Amber::Router
 
     property route_set : RouteSet(T)
     property segment : String
+    property requirement : Regex?
```
```suggestion
    property pattern : Regex?
```

Seems more clear about its nature.

Sija commented on src/amber/router/route_set.cr:
```diff
@@ -60,16 +61,16 @@ module Amber::Router
 
     # Add a route to the tree.
-    def add(path, payload : T) : Nil
+    def add(path, payload : T, requirements : Requirements = {} of String => Regex) : Nil
```
I'd add handling of requirements as `NamedTuple`.

Blacksmoke16 replied:
How would you know the keys, since they are mapped to the named params in the route.

Sija replied:
I'm not sure I understand. You can always transform it into `Hash` with `.to_h`.

Blacksmoke16 replied:
Ohh i see what you mean.  Yea i can prob do that.  Was thinking you wanted the type restriction to be `NamedTuple` and not a hash.  My bad.

Sija replied:
I belive this is still left.

robacarp reviewed (commented):
Some of the code paths are a bit vague in my memory, so I may be overlooking something obvious. Is it possible to encapsulate everything around the idea of a requirement to be in the VariableSegment class? Then the #match? method would take care of storing the regex, the logic of checking to see if the parameter matches the regex, etc. What do you think?

Blacksmoke16 committed:
Move logic into #match?

Add benchmark for route with requirements

Blacksmoke16 commented:
@robacarp Good call.  I added the pattern to only the variable segment class, and just handled the logic from within the `match?` method. 

I also added a section to the benchmark for a route with a requirement.  I should probably add one that fails to match now that i think of it... I'll do that when i get home.

@Sija I renamed the internal property to `pattern`, but left the param on the `add` method as `requirements` as it could include multiple patterns.

Sija commented:
Instead of `requirements` I'd suggest using `constraints`, which is more clear and commonly used term.

robacarp commented on src/amber/router/segments/variable_segment.cr:
```diff
@@ -1,7 +1,11 @@
+
     def match?(segment : String) : Bool
-      true
+      (p = @pattern) ? !(segment =~ p).nil? : true
```
This line is a bit hard to parse, can we make it a little more explicit?

```suggestion
      if pattern = @pattern
        segment =~ pattern
      else
        true
      end
```

Blacksmoke16 committed:
Change param name to constraint

Make the match logic easier to read
Add no match to benchmark

Blacksmoke16 commented:
@robacarp @Sija This should be good to go now.

drujensen commented:
@robacarp @Sija need your approval

Sija commented:
Just `NamedTuple` thing I mentioned in a comment before and it's GTG on my side.

Blacksmoke16 committed:
Support named tuple constraints

Blacksmoke16 committed:
Fix typo

Blacksmoke16 committed:
Revert ameba/format changes

Blacksmoke16 committed:
Remove this too

Blacksmoke16 commented:
Ok how about now.

Sija commented:
Why did you remove `ameba` dependency? Whatever the motivation you should open an issue for that since it's our of scope of this PR.

Blacksmoke16 commented:
It's implemented in #13

Sija commented:
Oh, I see, fair enough!

robacarp reviewed (approved):
Thanks @Blacksmoke16, this looks great.
